keras_retinanet/utils/eval.py

```diff
@@ -0,0 +1,192 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from __future__ import print_function
+
+from .anchors import compute_overlap
+from .visualization import draw_detections, draw_ground_truth
+
+import numpy as np
+import os
+
+import cv2
+import pickle
+
+
+def _compute_ap(recall, precision):
+    # code originally from https://github.com/rbgirshick/py-faster-rcnn
+
+    # correct AP calculation
+    # first append sentinel values at the end
+    mrec = np.concatenate(([0.], recall, [1.]))
+    mpre = np.concatenate(([0.], precision, [0.]))
+
+    # compute the precision envelope
+    for i in range(mpre.size - 1, 0, -1):
+        mpre[i - 1] = np.maximum(mpre[i - 1], mpre[i])
+
+    # to calculate area under PR curve, look for points
+    # where X axis (recall) changes value
+    i = np.where(mrec[1:] != mrec[:-1])[0]
+
+    # and sum (\Delta recall) * prec
+    ap = np.sum((mrec[i + 1] - mrec[i]) * mpre[i + 1])
+    return ap
+
+
+def _get_detections(generator, model, score_threshold=0.05, max_detections=100, save_path=None):
+    all_detections = [[None for i in range(generator.num_classes())] for j in range(generator.size())]
+
+    for i in range(generator.size()):
+        raw_image    = generator.load_image(i)
+        image        = generator.preprocess_image(raw_image.copy())
+        image, scale = generator.resize_image(image)
+
+        # run network
+        _, _, detections = model.predict_on_batch(np.expand_dims(image, axis=0))
+
+        # clip to image shape
+        detections[:, :, 0] = np.maximum(0, detections[:, :, 0])
+        detections[:, :, 1] = np.maximum(0, detections[:, :, 1])
+        detections[:, :, 2] = np.minimum(image.shape[1], detections[:, :, 2])
+        detections[:, :, 3] = np.minimum(image.shape[0], detections[:, :, 3])
+
+        # correct boxes for image scale
+        detections[0, :, :4] /= scale
+
+        # select scores from detections
+        scores = detections[0, :, 4:]
+
+        # select indices which have a score above the threshold
+        indices = np.where(detections[0, :, 4:] > score_threshold)
+
+        # select those scores
+        scores = scores[indices]
+
+        # find the order with which to sort the scores
+        scores_sort = np.argsort(-scores)[:max_detections]
+
+        # select detections
+        image_boxes      = detections[0, indices[0][scores_sort], :4]
+        image_scores     = np.expand_dims(detections[0, indices[0][scores_sort], 4 + indices[1][scores_sort]], axis=1)
+        image_detections = np.append(image_boxes, image_scores, axis=1)
+        image_predicted_labels = indices[1][scores_sort]
+
+        if save_path:
```

Comment:
I would go for `if save_path is not None:` here. Otherwise, `--save-path ''` on the command line doesn't save any images. I would expect the images to be saved to the current working directory. I would be a bit annoyed if it turns out afterwards that no images were saved.

`--save-path .` would be a workaround, but I don't see a reason to special case empty strings here.

Reply:
Agreed, edited.